scripts/west_commands/runners/stm32cubeprogrammer.py

```diff
@@ -0,0 +1,199 @@
+        use_elf: bool,
+        erase: bool,
+        extra_opts: List[str],
+    ) -> None:
+        super().__init__(cfg)
+
+        self._port = port
+        self._frequency = frequency
+        self._reset_mode = reset_mode
+        self._extra_connect_opts = extra_connect_opts
+        self._cli = (
+            cli or STM32CubeProgrammerBinaryRunner._get_stm32cubeprogrammer_path()
+        )
+        self._use_elf = use_elf
+        self._erase = erase
+
+        self._extra_opts: List[str] = list()
+        for opts in [shlex.split(opt) for opt in extra_opts]:
+            self._extra_opts += opts
+
+        # add required library loader path to the environment (Linux only)
+        if platform.system() == "Linux":
+            os.environ["LD_LIBRARY_PATH"] = str(self._cli.parent / ".." / "lib")
+
+    @staticmethod
+    def _get_stm32cubeprogrammer_path() -> Path:
+        """Obtain path of the STM32CubeProgrammer CLI tool."""
+
+        if platform.system() == "Linux":
+            return (
+                Path.home()
+                / "STMicroelectronics"
+                / "STM32Cube"
+                / "STM32CubeProgrammer"
+                / "bin"
+                / "STM32_Programmer_CLI"
+            )
+
+        if platform.system() == "Windows":
+            return (
+                Path(os.environ["PROGRAMFILES"])
+                / "STMicroelectronics"
+                / "STM32Cube"
+                / "STM32CubeProgrammer"
+                / "bin"
+                / "STM32_Programmer_CLI.exe"
+            )
+
+        if platform.system() == "Darwin":
+            return (
+                Path("/Applications")
+                / "STMicroelectronics"
+                / "STM32Cube"
+                / "STM32CubeProgrammer"
+                / "STM32CubeProgrammer.app"
+                / "Contents"
+                / "MacOs"
+                / "bin"
+                / "STM32_Programmer_CLI"
+            )
+
+        raise NotImplementedError("Could not determine STM32_Programmer_CLI path")
+
+    @classmethod
+    def name(cls):
+        return "stm32cubeprogrammer"
+
+    @classmethod
+    def capabilities(cls):
+        return RunnerCaps(commands={"flash"}, erase=True)
+
+    @classmethod
+    def do_add_parser(cls, parser):
+        parser.add_argument(
+            "--port",
+            type=str,
+            required=True,
+            help="Interface identifier, e.g. swd, jtag, /dev/ttyS0...",
+        )
+        parser.add_argument(
+            "--frequency", type=int, required=False, help="Programmer frequency in KHz"
+        )
+        parser.add_argument(
+            "--reset-mode",
+            type=str,
+            required=False,
+            choices=["sw", "hw", "core"],
+            help="Reset mode",
+        )
+        parser.add_argument(
+            "--extra-connect-opts",
+            type=str,
+            required=False,
+            help="Additional options for the --connect argument",
```

Comment:
We have various legacy options in different runners that map directly to an option that is passed to the underlying tool, but I am trying to get away from doing this by adding a single `--tool-opt` option that says "hey runner, here is a thing you can pass to the program you're a wrapper around".

The nrfjprog, jlink, and pyocd runners implement this today. I see this here and think "could this be a --tool-opt instead"?

Are there any other options that you think might have a "yes" answer to this question in this PR?

Reply:
I did not know `--tool-opt` was standardized for passing arbitrary tool options. Should be put in the base class as `--erase` then? The `connect-opts` will be needed, as these are appended options specific to the connection set.

Reply:
> I did not know `--tool-opt` was standardized for passing arbitrary tool options. Should it be put in the base class as `--erase` then?

Yeah, eventually it probably should be a common capability; I just haven't gotten around to it yet. I mostly just ask people for this change when I see this sort of option being added.

> The `connect-opts` will be needed, as these are appended options specific to the connection set.

OK, so kind of like --dfuse-modifiers then? Seems reasonable to me.

Reply:
@mbolivar fixed, used `--conn-modifiers`